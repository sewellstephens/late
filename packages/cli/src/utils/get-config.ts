import { cosmiconfig } from 'cosmiconfig';
import path from 'path';
import { loadConfig } from 'tsconfig-paths';
import * as z from 'zod';

import { resolveImport } from './resolve-import';

/**
 * This module is primarily concerned with loading and validating a project's
 * configuration. It makes use of `cosmiconfig` to find and load a configuration
 * file, in this case, it's looking for a `plate-components.json` or
 * `components.json` file in the root directory of your project. The schema for
 * this config file is defined with the `zod` library. In case there's no
 * configuration file available, it provides a set of default paths and
 * configuration options.
 */

export const DEFAULT_STYLE = 'default';

export const DEFAULT_PLATE_UI = '@/components/plate-ui';

export const DEFAULT_COMPONENTS = '@/components';

export const DEFAULT_TAILWIND_CSS = 'src/styles/globals.css';

export const DEFAULT_TAILWIND_CONFIG = 'tailwind.config.js';

export const DEFAULT_TAILWIND_BASE_COLOR = 'slate';

const explorerPlateComponents = cosmiconfig('components', {
  searchPlaces: ['plate-components.json'],
});
const explorerComponents = cosmiconfig('components', {
  searchPlaces: ['components.json'],
});

export const rawConfigSchema = z.object({
  $schema: z.string().optional(),
  aliases: z.object({
    components: z.string(),
    'plate-ui': z.string().optional(),
    ui: z.string().optional(),
  }),
  rsc: z.coerce.boolean().default(false),
  style: z.string(),
  tailwind: z.object({
    baseColor: z.string(),
    config: z.string(),
    css: z.string(),
    cssVariables: z.boolean().default(true),
    prefix: z.string().default('').optional(),
  }),
});

export type RawConfig = z.infer<typeof rawConfigSchema>;

export const configSchema = rawConfigSchema.extend({
  resolvedPaths: z.object({
    components: z.string(),
    'plate-ui': z.string(),
    tailwindConfig: z.string(),
    tailwindCss: z.string(),
    ui: z.string(),
  }),
});

export type Config = z.infer<typeof configSchema>;

export async function getConfig(cwd: string) {
  const config = await getRawConfig(cwd);

  if (!config) {
    return null;
  }

  return await resolveConfigPaths(cwd, config);
}

export async function resolveConfigPaths(cwd: string, config: RawConfig) {
  // Read tsconfig.json.
  const tsConfig = loadConfig(cwd);

  if (tsConfig.resultType === 'failed') {
    throw new Error(
      `Failed to load tsconfig.json. ${tsConfig.message ?? ''}`.trim()
    );
  }

  const ui = config.aliases.ui
    ? await resolveImport(config.aliases.ui, tsConfig)
    : await resolveImport(config.aliases.components, tsConfig);

  return configSchema.parse({
    ...config,
    resolvedPaths: {
      components: await resolveImport(config.aliases.components, tsConfig),
      'plate-ui': config.aliases['plate-ui']
        ? await resolveImport(config.aliases['plate-ui'], tsConfig)
        : ui,
      tailwindConfig: path.resolve(cwd, config.tailwind.config),
      tailwindCss: path.resolve(cwd, config.tailwind.css),
      ui: ui,
    },
  });
}

export async function getRawConfig(cwd: string): Promise<RawConfig | null> {
  let filename = 'plate-components';

  try {
    let configResult = await explorerPlateComponents.search(cwd);

    if (!configResult) {
      filename = 'components';
      configResult = await explorerComponents.search(cwd);

      if (!configResult) {
        return null;
      }
    }

    return rawConfigSchema.parse(configResult.config);
  } catch (error) {
    throw new Error(`Invalid configuration found in ${cwd}/${filename}.json.`);
  }
}
