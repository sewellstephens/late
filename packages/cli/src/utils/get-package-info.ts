// eslint-disable-next-line import/no-extraneous-dependencies
import type { PackageJson } from 'type-fest';

import fs from 'fs-extra';
import path from 'path';

/**
 * This is a simple utility that reads and returns the contents of your
 * project's package.json file.
 */
export function getPackageInfo() {
  const packageJsonPath = path.join('package.json');

  return fs.readJSONSync(packageJsonPath) as PackageJson;
}
