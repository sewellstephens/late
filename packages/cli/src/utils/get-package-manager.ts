import { detect } from '@antfu/ni';

/**
 * This script uses `@antfu/ni` to automatically detect the package manager used
 * in the current project (yarn, pnpm, or npm).
 */
export async function getPackageManager(
  targetDir: string
): Promise<'bun' | 'npm' | 'pnpm' | 'yarn'> {
  const packageManager = await detect({ cwd: targetDir, programmatic: true });

  if (packageManager === 'yarn@berry') return 'yarn';
  if (packageManager === 'pnpm@6') return 'pnpm';
  if (packageManager === 'bun') return 'bun';

  return packageManager ?? 'npm';
}
