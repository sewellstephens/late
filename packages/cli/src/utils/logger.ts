/* eslint-disable no-console */
import chalk from 'chalk';

/**
 * This is a simple logging utility, which wraps console.log and prefixes
 * messages with colored severity levels. This can be used throughout the
 * project to log informative, warning, error, and success messages in a
 * consistent way.
 */
export const logger = {
  break() {
    console.info('');
  },
  error(...args: unknown[]) {
    console.info(chalk.red(...args));
  },
  info(...args: unknown[]) {
    console.info(chalk.cyan(...args));
  },
  success(...args: unknown[]) {
    console.info(chalk.green(...args));
  },
  warn(...args: unknown[]) {
    console.info(chalk.yellow(...args));
  },
};
