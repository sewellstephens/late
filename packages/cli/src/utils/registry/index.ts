/* eslint-disable turbo/no-undeclared-env-vars,no-console */

import type * as z from 'zod';

import { HttpsProxyAgent } from 'https-proxy-agent';
import fetch from 'node-fetch';
import path from 'path';

import type { Config } from '../get-config';

import {
  registryBaseColorSchema,
  registryIndexSchema,
  type registryItemWithContentSchema,
  registryWithContentSchema,
  stylesSchema,
} from './schema';

const baseUrl = process.env.COMPONENTS_REGISTRY_URL ?? 'https://sewellstephens.github.io/late';
const agent = process.env.https_proxy
  ? new HttpsProxyAgent(process.env.https_proxy)
  : undefined;

export async function getRegistryIndex() {
  try {
    const [result] = await fetchRegistry(['index.json']);

    return registryIndexSchema.parse(result);
  } catch (error) {
    throw new Error(`Failed to fetch components from registry.`);
  }
}

export async function getRegistryStyles() {
  try {
    const [result] = await fetchRegistry(['styles/index.json']);

    return stylesSchema.parse(result);
  } catch (error) {
    throw new Error(`Failed to fetch styles from registry.`);
  }
}

// eslint-disable-next-line @typescript-eslint/require-await
export async function getRegistryBaseColors() {
  return [
    {
      label: 'Slate',
      name: 'slate',
    },
    {
      label: 'Gray',
      name: 'gray',
    },
    {
      label: 'Zinc',
      name: 'zinc',
    },
    {
      label: 'Neutral',
      name: 'neutral',
    },
    {
      label: 'Stone',
      name: 'stone',
    },
  ];
}

export async function getRegistryBaseColor(baseColor: string) {
  try {
    const [result] = await fetchRegistry([`colors/${baseColor}.json`]);

    return registryBaseColorSchema.parse(result);
  } catch (error) {
    throw new Error(`Failed to fetch base color from registry.`);
  }
}

export async function resolveTree(
  index: z.infer<typeof registryIndexSchema>,
  names: string[]
) {
  const tree: z.infer<typeof registryIndexSchema> = [];

  for (const name of names) {
    const entry = index.find((e) => e.name === name);

    if (!entry) {
      continue;
    }

    tree.push(entry);

    if (entry.registryDependencies) {
      const dependencies = await resolveTree(index, entry.registryDependencies);
      tree.push(...dependencies);
    }
  }

  return tree.filter(
    (component, i, self) =>
      self.findIndex((c) => c.name === component.name) === i
  );
}

export async function fetchTree(
  style: string,
  tree: z.infer<typeof registryIndexSchema>
) {
  try {
    const paths = tree.map((item) => `styles/${style}/${item.name}.json`);
    const result = await fetchRegistry(paths);

    return registryWithContentSchema.parse(result);
  } catch (error) {
    throw new Error(`Failed to fetch tree from registry.`);
  }
}

// eslint-disable-next-line @typescript-eslint/require-await
export async function getItemTargetPath(
  config: Config,
  item: Pick<z.infer<typeof registryItemWithContentSchema>, 'type'>,
  override?: string
) {
  if (override) {
    return override;
  }

  if (item.type === 'components:plate-ui' && config?.aliases['plate-ui']) {
    return config.resolvedPaths['plate-ui'];
  }

  if (item.type === 'components:plate-ui' && config?.aliases.ui) {
    return config.resolvedPaths.ui;
  }

  const [parent, type] = item.type.split(':');
  if (!config || !(parent in config.resolvedPaths)) {
    return null;
  }

  return path.join(
    config.resolvedPaths[parent as keyof typeof config.resolvedPaths],
    type
  );
}

async function fetchRegistry(paths: string[]) {
  try {
    const results = await Promise.all(
      paths.map(async (p) => {
        const response = await fetch(`${baseUrl}/registry/${p}`, {
          agent,
        });
        return await response.json();
      })
    );

    return results;
  } catch (error) {
    console.error(error);
    throw new Error(`Failed to fetch registry from ${baseUrl}.`);
  }
}
