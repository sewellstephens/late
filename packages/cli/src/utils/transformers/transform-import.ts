import type { Transformer } from '.';

// eslint-disable-next-line @typescript-eslint/require-await
export const transformImport: Transformer = async ({ config, sourceFile }) => {
  const importDeclarations = sourceFile.getImportDeclarations();

  for (const importDeclaration of importDeclarations) {
    const moduleSpecifier = importDeclaration.getModuleSpecifierValue();

    // Replace @/registry/[style] with the components alias.
    if (moduleSpecifier.startsWith('@/registry/')) {
      if (config.aliases['plate-ui']) {
        importDeclaration.setModuleSpecifier(
          moduleSpecifier.replace(
            /^@\/registry\/[^/]+\/plate-ui/,
            config.aliases['plate-ui']
          )
        );
      } else if (config.aliases.ui) {
        importDeclaration.setModuleSpecifier(
          moduleSpecifier.replace(
            /^@\/registry\/[^/]+\/plate-ui/,
            config.aliases.ui
          )
        );
      } else {
        importDeclaration.setModuleSpecifier(
          moduleSpecifier.replace(
            /^@\/registry\/[^/]+/,
            config.aliases.components
          )
        );
      }
    }
  }

  return sourceFile;
};
