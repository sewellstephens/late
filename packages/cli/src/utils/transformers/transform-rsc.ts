import { SyntaxKind } from 'ts-morph';

import type { Transformer } from '.';

// eslint-disable-next-line @typescript-eslint/require-await
export const transformRsc: Transformer = async ({ config, sourceFile }) => {
  if (config.rsc) {
    return sourceFile;
  }

  // Remove "use client" from the top of the file.
  const first = sourceFile.getFirstChildByKind(SyntaxKind.ExpressionStatement);

  if (first?.getText() === `'use client'`) {
    first.remove();
  }

  return sourceFile;
};
