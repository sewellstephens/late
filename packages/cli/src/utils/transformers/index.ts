import type { SourceFile } from 'ts-morph';
import type * as z from 'zod';

import { promises as fs } from 'fs';
import { tmpdir } from 'os';
import path from 'path';
import { Project, QuoteKind, ScriptKind } from 'ts-morph';

import type { Config } from '../get-config';
import type { registryBaseColorSchema } from '../registry/schema';

import { transformCssVars } from './transform-css-vars';
import { transformImport } from './transform-import';
import { transformRsc } from './transform-rsc';
import { transformTwPrefixes } from './transform-tw-prefix';

export type TransformOpts = {
  baseColor?: z.infer<typeof registryBaseColorSchema>;
  config: Config;
  filename: string;
  raw: string;
};

export type Transformer = (
  opts: {
    sourceFile: SourceFile;
  } & TransformOpts
) => Promise<SourceFile>;

const transformers: Transformer[] = [
  transformImport,
  transformRsc,
  transformCssVars,
  transformTwPrefixes,
];

const project = new Project({
  compilerOptions: {},
  manipulationSettings: {
    quoteKind: QuoteKind.Single,
  },
});

async function createTempSourceFile(filename: string) {
  const dir = await fs.mkdtemp(path.join(tmpdir(), 'plate-'));
  return path.join(dir, filename);
}

export async function transform(opts: TransformOpts) {
  const tempFile = await createTempSourceFile(opts.filename);
  const sourceFile = project.createSourceFile(tempFile, opts.raw, {
    scriptKind: ScriptKind.TSX,
  });

  for (const transformer of transformers) {
    void transformer({ sourceFile, ...opts });
  }

  return sourceFile.getFullText();
}
