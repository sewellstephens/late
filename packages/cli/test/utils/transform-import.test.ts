import { expect, test } from 'vitest';

import { transform } from '../../src/utils/transformers';

test('transform import', async () => {
  expect(
    await transform({
      config: {
        aliases: {
          components: '@/components',
        },
        tailwind: {
          baseColor: 'neutral',
          cssVariables: true,
        },
      } as any,
      filename: 'test.ts',
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/default/plate-ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/default/box"

    import { cn } from "@sewell_stephens/cn"
    `,
    })
  ).toMatchSnapshot();

  expect(
    await transform({
      config: {
        aliases: {
          components: '~/src/components',
        },
      } as any,
      filename: 'test.ts',
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/default/plate-ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/default/box"

    import { cn, foo, bar } from "@sewell_stephens/cn"
    `,
    })
  ).toMatchSnapshot();

  expect(
    await transform({
      config: {
        aliases: {
          components: '~/src/components',
        },
      } as any,
      filename: 'test.ts',
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/default/plate-ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/default/box"

    import { cn } from "@sewell_stephens/cn"
    `,
    })
  ).toMatchSnapshot();

  expect(
    await transform({
      config: {
        aliases: {
          components: '~/src/components',
          'plate-ui': '~/src/components',
        },
      } as any,
      filename: 'test.ts',
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/default/plate-ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/default/box"
    import { cn } from "@/lib/utils"
    import { bar } from "@/lib/utils/bar"
    `,
    })
  ).toMatchSnapshot();

  expect(
    await transform({
      config: {
        aliases: {
          components: '~/src/components',
          'plate-ui': '~/src/plate-ui',
        },
      } as any,
      filename: 'test.ts',
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/default/plate-ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/default/box"
    import { cn } from "@/lib/utils"
    import { bar } from "@/lib/utils/bar"
    `,
    })
  ).toMatchSnapshot();

  expect(
    await transform({
      config: {
        aliases: {
          components: '~/src/components',
          ui: '~/src/components',
        },
      } as any,
      filename: 'test.ts',
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/default/plate-ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/default/box"
    import { cn } from "@/lib/utils"
    import { bar } from "@/lib/utils/bar"
    `,
    })
  ).toMatchSnapshot();

  expect(
    await transform({
      config: {
        aliases: {
          components: '~/src/components',
          ui: '~/src/ui',
        },
      } as any,
      filename: 'test.ts',
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/default/plate-ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/default/box"
    import { cn } from "@/lib/utils"
    import { bar } from "@/lib/utils/bar"
    `,
    })
  ).toMatchSnapshot();
});
