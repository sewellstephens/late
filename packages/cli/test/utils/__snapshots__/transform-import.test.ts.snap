// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform import 1`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"@/components/plate-ui/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"@/components/box\\"

    import { cn } from \\"@sewell_stephens/cn\\"
    "
`;

exports[`transform import 2`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"~/src/components/plate-ui/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"~/src/components/box\\"

    import { cn, foo, bar } from \\"@sewell_stephens/cn\\"
    "
`;

exports[`transform import 3`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"~/src/components/plate-ui/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"~/src/components/box\\"

    import { cn } from \\"@sewell_stephens/cn\\"
    "
`;

exports[`transform import 4`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"~/src/components/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"@/registry/default/box\\"
    import { cn } from \\"@/lib/utils\\"
    import { bar } from \\"@/lib/utils/bar\\"
    "
`;

exports[`transform import 5`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"~/src/plate-ui/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"@/registry/default/box\\"
    import { cn } from \\"@/lib/utils\\"
    import { bar } from \\"@/lib/utils/bar\\"
    "
`;

exports[`transform import 6`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"~/src/components/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"@/registry/default/box\\"
    import { cn } from \\"@/lib/utils\\"
    import { bar } from \\"@/lib/utils/bar\\"
    "
`;

exports[`transform import 7`] = `
"import * as React from \\"react\\"
import { Foo } from \\"bar\\"
    import { Button } from \\"~/src/ui/button\\"
    import { Label} from \\"ui/label\\"
    import { Box } from \\"@/registry/default/box\\"
    import { cn } from \\"@/lib/utils\\"
    import { bar } from \\"@/lib/utils/bar\\"
    "
`;
